eval/topdown.go

```diff
@@ -489,18 +563,55 @@ func evalRefRec(ctx *TopDownContext, ref opalog.Ref, iter TopDownIterator, path
 
 	headVar, isVar := head.Value.(opalog.Var)
 
-	if !isVar || len(path) == 0 {
+	// Handle head of reference.
+	if isVar && len(path) == 0 {
+
+		// Check if head has a binding. E.g., x = [1,2,3], x[i] = 1
+		binding := ctx.Bindings.Get(headVar)
+		if binding != nil {
+			return evalRefRuleResult(ctx, ref, ref[1:], binding, iter)
+		}
+
 		path = append(path, head)
-		return evalRefRec(ctx, tail, iter, path)
+		node, err := ctx.Store.Lookup(path)
+		if err != nil {
+			switch err := err.(type) {
+			case *StorageError:
+				if err.Code == StorageNotFoundErr {
+					return nil
+				}
+			}
+			return err
+		}
+
+		switch node := node.(type) {
+		case []*opalog.Rule:
+			for _, rule := range node {
+				if err := evalRefRule(ctx, ref, path, rule, iter); err != nil {
+					return err
+				}
+			}
+			return nil
+		default:
+			return evalRefRec(ctx, tail, iter, path)
+		}
 	}
 
-	binding, isBound := ctx.Bindings[headVar]
+	// Handle constants in reference.
+	if !isVar {
+		path = append(path, head)
+		return evalRefRec(ctx, tail, iter, path)
+	}
 
-	if isBound {
+	// Binding already exists for variable. Treat it as a constant.
+	binding := ctx.Bindings.Get(headVar)
```

Comment:
Sorry, this is a bit confusing. `evalRefRec` could probably be split up a bit. I will see what I can do.

The case that I think you're referring to is handled above. E.g., `x = [1,2,3,4], x[i] = 2`. When the second expression is evaluated, `x` is bound to `[1,2,3,4]`. This is handled on line 572.

The line you commented on is dealing with the case where you have some reference, e.g., `a.b[i].c[j]` where the variables in the tail of the reference already have bindings. The variable `headVar` is a bit confusing in this context. The reference is being processed recursively so the head is not necessarily the very beginning of the reference.